src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java

```diff
@@ -421,47 +421,55 @@ void writeTo(BufWriterImpl pool) {
                 }
                 pool.writeU1(tag);
                 pool.writeU2(charLen);
-                for (int i = 0; i < charLen; ++i) {
-                    char c = stringValue.charAt(i);
-                    if (c >= '\001' && c <= '\177') {
-                        // Optimistic writing -- hope everything is bytes
-                        // If not, we bail out, and alternate path patches the length
-                        pool.writeU1((byte) c);
-                    }
-                    else {
-                        int charLength = stringValue.length();
-                        int byteLength = i;
-                        char c1;
-                        for (int j = i; j < charLength; ++j) {
-                            c1 = (stringValue).charAt(j);
-                            if (c1 >= '\001' && c1 <= '\177') {
-                                byteLength++;
-                            } else if (c1 > '\u07FF') {
-                                byteLength += 3;
-                            } else {
-                                byteLength += 2;
-                            }
-                        }
-                        if (byteLength > 65535) {
-                            throw new IllegalArgumentException();
+                if (!JLA.hasNegativeOrZeros(stringValue)) {
+                    pool.writeBytesDirect(stringValue);
+                } else {
+                    writeToSlow(pool);
+                }
+            }
+        }
+
+        private void writeToSlow(BufWriterImpl pool) {
+            for (int i = 0; i < charLen; ++i) {
+                char c = stringValue.charAt(i);
+                if (c >= '\001' && c <= '\177') {
+                    // Optimistic writing -- hope everything is bytes
+                    // If not, we bail out, and alternate path patches the length
+                    pool.writeU1((byte) c);
+                }
+                else {
+                    int charLength = stringValue.length();
```

Comment:
In the slow loop, you can remove everything before the `else` - we don't really need that optimistic writing any more

Reply:
The code of the slow path is complex and difficult to understand. Can the simplified work be used as a follow-up action?

Reply:
Sure! Good to focus on tge improvement here.